dbms/src/DataStreams/WindowBlockInputStream.cpp

```diff
@@ -0,0 +1,574 @@
+    {
+        // For RANGE frames, rows that compare equal w/ORDER BY are peers.
+        assert(window_description.frame.type == WindowFrame::FrameType::Ranges);
+        const size_t n = order_column_indices.size();
+        if (n == 0)
+        {
+            // No ORDER BY, so all rows are peers.
+            return true;
+        }
+
+        for (size_t i = 0; i < n; ++i)
+        {
+            const auto * column_x = inputAt(x)[order_column_indices[i]].get();
+            const auto * column_y = inputAt(y)[order_column_indices[i]].get();
+            if (window_description.order_by[i].collator)
+            {
+                if (column_x->compareAtWithCollation(x.row, y.row, *column_y, 1 /* nan_direction_hint */, *window_description.order_by[i].collator) != 0)
+                {
+                    return false;
+                }
+            }
+            else
+            {
+                if (column_x->compareAt(x.row, y.row, *column_y, 1 /* nan_direction_hint */) != 0)
+                {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+    case WindowFrame::FrameType::Groups:
+    default:
+        throw Exception(ErrorCodes::NOT_IMPLEMENTED,
+                        "window function only support frame type row and range.");
+    }
+}
+
+void WindowBlockInputStream::advanceFrameEndCurrentRow()
+{
+    assert(frame_end.block == partition_end.block
+           || frame_end.block + 1 == partition_end.block);
+
+    // If window only have row_number or rank/dense_rank functions, set frame_end to the next row of current_row and frame_ended to true
+    if (only_have_pure_window)
+    {
+        frame_end = current_row;
+        advanceRowNumber(frame_end);
+        frame_ended = true;
+        return;
+    }
+
+    throw Exception(ErrorCodes::NOT_IMPLEMENTED,
+                    "window function only support pure window function now.");
+}
+
+void WindowBlockInputStream::advanceFrameEnd()
+{
+    // No reason for this function to be called again after it succeeded.
+    assert(!frame_ended);
+
+    // switch for another frame type
+    switch (window_description.frame.end_type)
+    {
+    case WindowFrame::BoundaryType::Current:
+        advanceFrameEndCurrentRow();
+        break;
+    case WindowFrame::BoundaryType::Unbounded:
+    case WindowFrame::BoundaryType::Offset:
+    default:
+        throw Exception(ErrorCodes::NOT_IMPLEMENTED,
+                        "The frame end type '{}' is not implemented",
+                        window_description.frame.end_type);
+    }
+}
+
+void WindowBlockInputStream::writeOutCurrentRow()
+{
+    assert(current_row < partition_end);
+    assert(current_row.block >= first_block_number);
+
+    for (size_t wi = 0; wi < workspaces.size(); ++wi)
+    {
+        auto & ws = workspaces[wi];
+        ws.window_function->windowInsertResultInto(this->shared_from_this(), wi);
+    }
+}
+
+Block WindowBlockInputStream::getOutputBlock()
```

Comment:
rename to `tryGetOutputBlock`?
Because you may not get the output block

Reply:
> Is it possible to separate the logic of the window function calculation into a new function? like
> 
> ```
> appendBlock(current_block);
> tryCalculate(); // other names are ok
> ```

done.